### PR TITLE
Canvass: select assigned area

### PR DESCRIPTION
**Pull request by rebecarubio**
## Description
This PR introduces an intermediate step between the canvassing instructions and the canvass map, allowing users to select from their assigned areas. If the user is only assigned to one area the page will redirect it to the map automatically. 
It also loads only the locations of the area selected and removes old code.

## Screenshots
Select area step:
image

Area selected in map:
image

## Changes
* Adds new `CanvassSelectAreaPage` component.
* Changes `/map` route to `/area` and adds the area id in the url.
* Changes `useLocations()` hook to load only the locations of the area sent in the parameters.
* Adds a new `locationsByAssignmentIdAndAreaId` list in the store and updates it in all location reducers.
* Adds back button to `CanvassPage`.
* Refactors `GLCanvassMap` component to work with only the selected area.
* Removes dead code.

## Notes to reviewer
* Play around with selecting areas, creating locations and visits
* Height is hardcoded in `ListItemButton` in `CanvassSelectAreaPage` to fixed row height, meaning same space for rows with or without area description.

## Related issues
None

### CODE REVIEW

File: src/app/canvass/[areaAssId]/area/[areaId]/page.tsx [hunk 1]
```diff
@@ -22,8 +23,8 @@ export default async function Page({ params }: PageProps) {
   try {
     await apiClient.get<ZetkinOrganization>(`/api/users/me`);
 
-    return <CanvassPage areaAssId={areaAssId} />;
+    return <CanvassPage areaAssId={areaAssId} areaId={areaId} />;
   } catch (err) {
-    return redirect(`/login?redirect=/canvass/${areaAssId}/map`);
+    return redirect(`/login?redirect=/canvass/${areaAssId}/area`);
```
Comment (richardolsson) [inline src/app/canvass/[areaAssId]/area/[areaId]/page.tsx:28]: I think that the URL should be `areas` (plural). Also, shouldn't this URL contain the `areaId`, so that the user comes back to the exact URL after logging in?

File: src/app/canvass/[areaAssId]/area/page.tsx [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import 'leaflet/dist/leaflet.css';
```
Comment (richardolsson) [inline src/app/canvass/[areaAssId]/area/page.tsx:1]: What is the purpose of importing the `leaflet` CSS file? I believe that this may not be necessary since we migrated away from leaflet, and worry that it could even cause problems. If this is result from copying a pattern on some other page (e.g. the canvassing map page?), it might even be best to remove it there.

File: src/features/canvass/components/GLCanvassMap/index.tsx [hunk 3]
```diff
@@ -57,7 +57,7 @@ const GLCanvassMap: FC<Props> = ({ areas, assignment }) => {
       [180, 90],
     ];
 
-    const areaHoles = areas.map((area) => area.boundary.coordinates[0]);
+    const areaHoles = [selectedArea.boundary.coordinates[0]];
```
Comment (richardolsson) [inline src/features/canvass/components/GLCanvassMap/index.tsx:60]: I guess it's no longer necessary to use an array for this. Note how the array is just spread on line 64, so the contents of the array (a single item) could be moved there instead to simplify the code.

File: src/features/canvass/components/CanvassPage.tsx [hunk 4]
```diff
@@ -51,30 +57,42 @@ const Page: FC<{ assignment: ZetkinAreaAssignment }> = ({ assignment }) => {
...
+              py={2}
             >
-              <Box>
-                <Box display="flex" flexDirection="column">
+              <Box display="flex">
+                <ArrowLeftIcon
+                  fontSize="large"
+                  onClick={() => router.push(`/canvass/${assignment.id}/area`)}
+                  sx={{ alignSelf: 'center', cursor: 'pointer', mr: 1 }}
+                />
```
Comment (richardolsson) [inline src/features/canvass/components/CanvassPage.tsx:68]: I liked this interaction when I tried it, and it made me miss having the same type of button on the list page.

File: src/features/canvass/components/CanvassSelectAreaPage.tsx [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        >
+          <Box
+            alignItems="center"
+            bgcolor={oldTheme.palette.background.paper}
+            display="flex"
+            justifyContent="space-between"
+            padding={2}
+          >
+            <Box>
+              <Typography variant="body1">{assignment.title}</Typography>
```
Comment (richardolsson) [inline src/features/canvass/components/CanvassSelectAreaPage.tsx:53]: I would like a back button here, to go back to the assignment instructions etc. I don't know if this is something you have already discussed and maybe there's a reason not to do it?

File: src/features/canvass/components/CanvassSelectAreaPage.tsx [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+              <List disablePadding>
+                {areas.map((area) => (
+                  <React.Fragment key={area.id}>
+                    <ListItem key={area.id} disablePadding>
+                      <ListItemButton
+                        onClick={async () => {
+                          setLoadingAreaId(area.id);
+                          await router.push(
+                            `/canvass/${assignment.id}/area/${area.id}`
+                          );
```
Comment (richardolsson) [inline src/features/canvass/components/CanvassSelectAreaPage.tsx:75]: I think we should generally avoid using `router.push()` in cases where we can use `href` instead. That way, the browser will understand that something is a link, and can give the user option to open in separate tabs etc.

In this case, I expect `ListItemButton` to have a `href` property (as is often the case in MUI) but I could be wrong.

File: src/features/canvass/components/CanvassSelectAreaPage.tsx [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+                            <Typography color="text.secondary" variant="body2">
+                              {area.description}
+                            </Typography>
+                          )}
+                        </Box>
+                        {loadingAreaId === area.id ? (
+                          <CircularProgress size={20} />
+                        ) : (
+                          <ArrowForwardIos fontSize="small" />
+                        )}
```
Comment (richardolsson) [inline src/features/canvass/components/CanvassSelectAreaPage.tsx:106]: Very nice! 💯

File: src/features/canvass/components/CanvassInstructionsPage.tsx [hunk 8]
```diff
@@ -105,13 +107,17 @@ const Page: FC<{
...
+                  ? router.push(`/canvass/${assignment.id}/area`)
+                  : router.push(`/canvass/${assignment.id}/area/${areas[0].id}`)
+              }
               sx={{
                 width: '50%',
               }}
               variant="contained"
             >
-              <Msg id={messageIds.instructions.start} />
+              {areas.length > 1 ? 'Select Area' : 'Begin assignment'}
```
Comment (richardolsson) [inline src/features/canvass/components/CanvassInstructionsPage.tsx:120]: This should be internationalized.

File: src/features/canvass/components/CanvassInstructionsPage.tsx [hunk 9]
```diff
@@ -105,13 +107,17 @@ const Page: FC<{
           >
             <Button
               fullWidth
-              onClick={() => router.push(`/canvass/${assignment.id}/map`)}
+              onClick={() =>
+                areas.length > 1
+                  ? router.push(`/canvass/${assignment.id}/area`)
+                  : router.push(`/canvass/${assignment.id}/area/${areas[0].id}`)
+              }
```
Comment (richardolsson) [inline src/features/canvass/components/CanvassInstructionsPage.tsx:114]: Could we use `href` instead here?

File: src/features/canvass/components/CanvassInstructionsPage.tsx [hunk 10]
```diff
@@ -105,13 +107,17 @@ const Page: FC<{
           >
             <Button
               fullWidth
-              onClick={() => router.push(`/canvass/${assignment.id}/map`)}
+              onClick={() =>
+                areas.length > 1
```
Comment (richardolsson) [inline src/features/canvass/components/CanvassInstructionsPage.tsx:111]: Since this condition is used twice, I think it would make sense to give it a name, e.g. `const userMustSelectArea = areas.length > 1`.

File: src/features/areaAssignments/store.ts [hunk 11]
```diff
@@ -250,32 +255,55 @@ const areaAssignmentSlice = createSlice({
...
+      Object.keys(state.locationsByAssignmentIdAndAreaId).forEach((key) => {
+        const [keyAssignmentIdStr] = key.split(':');
+        const keyAssignmentId = Number(keyAssignmentIdStr);
+
+        if (keyAssignmentId == assignmentId) {
+          const list = state.locationsByAssignmentIdAndAreaId[key];
+          if (list.items.some((item) => item.id == location.id)) {
+            remoteItemUpdated(list, location);
+          }
+        }
```
Comment (richardolsson) [inline src/features/areaAssignments/store.ts:268]: Nice solution!

File: src/features/areaAssignments/store.ts [hunk 12]
```diff
@@ -250,32 +255,55 @@ const areaAssignmentSlice = createSlice({
...
     },
     locationUpdated: (state, action: PayloadAction<ZetkinLocation>) => {
       const location = action.payload;
 
       Object.values(state.locationsByAssignmentId).forEach((list) => {
         remoteItemUpdated(list, location);
       });
+      Object.values(state.locationsByAssignmentIdAndAreaId).forEach((list) => {
+        remoteItemUpdated(list, location);
+      });
```
Comment (richardolsson) [inline src/features/areaAssignments/store.ts:279]: I worry that overwriting the location object for all areas and assignments could overwrite stats from one assignment in the store for another. This is an edge case and is solved by refreshing the browser, so probably not a big deal. But it's something to think about and potentially revisit this for a refactor (and/or adding tests) down the line.

**Review by richardolsson — CHANGES_REQUESTED** [review]
Great work! I think this solves several bugs caused by having multiple areas in the map at the same time, and the UI feels very polished for an MVP. I put some comments below that can be improved and I'll let you decide what feels important enough to prioritize before merging. Whatever doesn't make it into this PR can be fixed later.

File: src/features/canvass/components/CanvassSelectAreaPage.tsx (hunk 5 above)
Comment (rebecarubio) [reply src/features/canvass/components/CanvassSelectAreaPage.tsx:53]: We didn't think about it, but I agree it’s a good idea to implement it to ensure smooth navigation between pages when needed.

File: src/features/canvass/components/CanvassSelectAreaPage.tsx (hunk 6 above)
Comment (rebecarubio) [reply src/features/canvass/components/CanvassSelectAreaPage.tsx:75]: It has a `href` property and it's working good. But I'm curious about why don't use the <Link> component that NextJS has?

File: src/app/canvass/[areaAssId]/area/page.tsx (hunk 2 above)
Comment (richardolsson) [reply src/app/canvass/[areaAssId]/area/page.tsx]: Fixed in 9b6a4ed6d4736a0619caa88296a439af1c114167.

File: src/features/canvass/components/GLCanvassMap/index.tsx (hunk 3 above)
Comment (richardolsson) [reply src/features/canvass/components/GLCanvassMap/index.tsx]: Fixed by 8d1265ea9885b8e524935ab4ecac170e9809e3d8.

File: src/features/canvass/components/CanvassInstructionsPage.tsx (hunk 10 above)
Comment (richardolsson) [reply src/features/canvass/components/CanvassInstructionsPage.tsx]: Fixed in bee5625a01f5f6d2dc7675bea5181cdd24c1c2bc

File: src/features/canvass/components/CanvassInstructionsPage.tsx (hunk 8 above)
Comment (richardolsson) [reply src/features/canvass/components/CanvassInstructionsPage.tsx]: Fixed in bee5625a01f5f6d2dc7675bea5181cdd24c1c2bc
